modal/_utils/async_utils.py

```diff
@@ -544,3 +553,117 @@ async def producer(producer_id: int, iterable: Union[AsyncIterable[T], Iterable[
 
 async def callable_to_agen(awaitable: Callable[[], Awaitable[T]]) -> AsyncGenerator[T, None]:
     yield await awaitable()
+
+
+async def async_map(
+    input_iterable: Union[AsyncIterable[T], Iterable[T]],
+    async_mapper_func: Callable[[T], Awaitable[V]],
+    concurrency: int,
+) -> AsyncGenerator[V, None]:
+    input_queue: asyncio.Queue[Tuple[str, Union[T, None]]] = asyncio.Queue()
+    output_queue: asyncio.Queue[Tuple[str, Union[V, Exception]]] = asyncio.Queue()
+    output_event = asyncio.Event()
+
+    async def producer():
+        async for item in sync_or_async_iter(input_iterable):
+            await input_queue.put((AsyncItemType.VALUE, item))
+        await input_queue.put((AsyncItemType.STOP, None))
+
+    async def worker():
+        while True:
+            try:
+                event_type, item = await input_queue.get()
+                if event_type == AsyncItemType.STOP:
+                    break
+                if event_type == AsyncItemType.VALUE:
+                    res = async_mapper_func(item)
+                    if inspect.isawaitable(res):
```

Comment:
I changed my mind - I think we should always assume `res` is awaitable and never allow blocking functions to be passed in here or to ordered_map_async - that would clean up a bunch of the code here, and I think we use async mapper functions in most cases anyways?